python-sdk/nuimages/scripts/render_images.py

```diff
@@ -64,30 +89,90 @@ def render_images(self,
+        print('Rendering %s for mode %s to folder %s...' % (out_type, mode, out_dir))
         for sample_token in tqdm.tqdm(sample_tokens):
             sample = self.nuim.get('sample', sample_token)
             sd_token_camera = sample['key_camera_token']
             sensor = self.nuim.shortcut('sample_data', 'sensor', sd_token_camera)
             sample_cam_name = sensor['channel']
+            sd_tokens_camera = self.nuim.get_sample_content(sample_token, modality='camera')
+
+            # We cannot render a video if there are missing camera sample_datas.
+            if len(sd_tokens_camera) < 13 and out_type == 'video':
+                continue
 
             for mode in modes:
-                out_path = os.path.join(out_dir, '%s_%s_%s.jpg' % (sample_token, sample_cam_name, mode))
-                if mode == 'annotated':
-                    self.nuim.render_image(sd_token_camera, with_annotations=True, out_path=out_path)
-                elif mode == 'image':
-                    self.nuim.render_image(sd_token_camera, with_annotations=False, out_path=out_path)
-                elif mode == 'depth_dense':
-                    self.nuim.render_depth(sd_token_camera, mode='dense', out_path=out_path)
-                elif mode == 'depth_sparse':
-                    self.nuim.render_depth(sd_token_camera, mode='sparse', out_path=out_path)
-                elif mode == 'pointcloud':
-                    sd_token_lidar = sample['key_lidar_token']
-                    self.nuim.render_pointcloud(sd_token_lidar, out_path=out_path)
-                elif mode == 'trajectory':
-                    self.nuim.render_trajectory(sample_token, out_path=out_path)
+                out_path_prefix = os.path.join(out_dir, '%s_%s_%s' % (sample_token, sample_cam_name, mode))
+                if out_type == 'image':
+                    self.write_image(sd_token_camera, mode, '%s.jpg' % out_path_prefix)
+                elif out_type == 'video':
+                    self.write_video(sd_tokens_camera, mode, out_path_prefix, cleanup=cleanup)
+
+    def write_video(self,
+                    sd_tokens_camera: List[str],
+                    mode: str,
+                    out_path_prefix: str,
+                    cleanup: bool = True) -> None:
+        """
+        Render a video by combining all the images of type mode for each sample_data.
+        :param sd_tokens_camera: All camera sample_data tokens in chronological order.
+        :param mode: The mode - see render_images().
+        :param out_path_prefix: The file prefix used for the images and video.
+        :param cleanup: Whether to delete images after rendering the video.
+        """
+        # Loop through each frame to create the video.
+        out_paths = []
+        for i, sd_token_camera in enumerate(sd_tokens_camera):
+            out_path = '%s_%d.jpg' % (out_path_prefix, i)
+            out_paths.append(out_path)
+            self.write_image(sd_token_camera, mode, out_path)
```

Comment:
Out of curiosity, why write the various frames to images first before combining them to form a video? Why not just write to a video directly (like in nuScenes)?

Reply:
It's tricky. We usually save via matplotlib's savefig(), as we don't actually have the final image. The only workaround I see is specifying a buffer instead of a file: https://stackoverflow.com/a/5314808

Reply:
Let's put it in our long-term todo list.

Reply:
Ahhh, icic; if you are going from plt to cv2, you can consider [this](https://github.com/nutonomy/nuscenes-devkit/blob/db6805fdc1336299a51e4c535929bd8714c87aca/python-sdk/nuscenes/lidarseg/lidarseg_utils.py#L44) next time; I had difficulty rendering vids for lidarseg previously since the scatter plot came from plt